source/dotnet/Library/AdaptiveCards/AdaptiveCard.cs

```diff
@@ -93,14 +121,47 @@ public static AdaptiveCardParseResult FromJson(string json)
         [JsonProperty(Order = -7, NullValueHandling = NullValueHandling.Ignore)]
         public string FallbackText { get; set; }
 
-        public bool ShouldSerializeActions()
+        /// <summary>
+        ///  Serialize this Adaptive Card to JSON
+        /// </summary>
+        /// <returns></returns>
+        public string ToJson()
+        {
+            return JsonConvert.SerializeObject(this, Formatting.Indented);
+        }
+    }
+
+    public class AdaptiveCardConverter : JsonConverter
+    {
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
         {
-            return Actions.Any();
+            throw new NotImplementedException();
         }
 
-        public string ToJson()
+        public override bool CanWrite => false;
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
         {
-            return JsonConvert.SerializeObject(this, Formatting.Indented);
+            var jObject = JObject.Load(reader);
+
+            var card = (AdaptiveCard) Activator.CreateInstance(objectType);
+            // Get the version from the payload
+            card.Version = null;
+
+            serializer.Populate(jObject.CreateReader(), card);
+
+            if (card.Version == null)
+                return null;
+
+            card.Actions.RemoveAll(item => item == null);
```

Comment:
Add a note here saying that unknown types come in as null? The only reason I vaguely know this is the case is because I overheard you saying that it parses unknown types as null or something... otherwise this code seems very unlogical and unexplained :)

Reply:
Also do we need the same logic when parsing containers/column sets? Do you have code there that does this logic? Or should we iterate through all Body descendants here and remove all null, not just for first-level children?